mempool/src/shared_mempool/peer_manager.rs

```diff
@@ -124,6 +181,161 @@ impl PeerManager {
+        // another peer, so don't track broadcasts for committed txns
+        state.broadcast_info.sent_batches = state
+            .broadcast_info
+            .sent_batches
+            .clone()
+            .into_iter()
+            .filter(|(id, _batch)| !mempool.timeline_range(id.0, id.1).is_empty())
+            .collect();
+
+        // check for batch to rebroadcast:
+        // 1. batch that did not receive ACK in configured window of time
+        // 2. batch that an earlier ACK marked as retriable
+        let rebroadcast = state
+            .broadcast_info
+            .sent_batches
+            .iter()
+            .find(|(_id, batch)| {
+                batch
+                    .send_time
+                    .checked_add(Duration::from_millis(
+                        self.mempool_config.shared_mempool_ack_timeout_ms,
+                    ))
+                    .map_or(false, |deadline| {
+                        SystemTime::now().duration_since(deadline).is_ok()
+                    })
+                    || matches!(batch.ack_status, AckStatus::Retry)
+            })
+            .map(|(id, _batch)| (*id, mempool.timeline_range(id.0, id.1)));
+
+        let (batch_id, transactions) = match rebroadcast {
+            Some((batch_id, txns)) => {
+                debug!("rebroadcasting {:?}", batch_id);
+                (batch_id, txns)
+            }
+            None => {
+                // fresh broadcast
+                let (txns, new_timeline_id) = mempool.read_timeline(
+                    state.timeline_id,
+                    self.mempool_config.shared_mempool_batch_size,
+                );
+                (BatchId(state.timeline_id, new_timeline_id), txns)
+            }
+        };
+
+        if transactions.is_empty() {
+            return;
+        }
+
+        // don't hold mempool lock during network send
+        drop(mempool);
+
+        // execute actual network send
+        let mut network_sender = smp
+            .network_senders
+            .get_mut(&peer.network_id())
+            .expect("[shared mempool] missing network sender")
+            .clone();
+        let num_txns = transactions.len();
+        let send_time = SystemTime::now();
+        if let Err(e) = tasks::send_mempool_sync_msg(
+            MempoolSyncMsg::BroadcastTransactionsRequest {
+                request_id: batch_id.create_request_id(),
+                transactions,
+            },
+            peer.peer_id(),
+            &mut network_sender,
+        ) {
+            error!(
+                "[shared mempool] error broadcasting transactions to peer {:?}: {}",
+                peer, e
+            );
+        } else {
+            let peer_id = &peer.peer_id().to_string();
+            counters::SHARED_MEMPOOL_BROADCAST_LATENCY
+                .with_label_values(&[peer_id])
+                .observe(start_time.elapsed().as_secs_f64());
```

Comment:
have you considered using... SHARED_MEMPOOL_BROADCAST_LATENCY::start_timer function on this instead? Also it would seem all this timer is doing is evaluating how quickly you can acquire a lock and generate a message to send, is that correct?

Reply:
I have this counter down here instead of using start_timer because I only wanted to record latency of actually executed broadcasts. and yes, this latency metrics is to measure the time to acquire a lock and execute to the network send. If this is too trivial I can remove it, I just have it here as a sanity check :P